### PR TITLE
nixos-container: Use new configuration & state directories

### LINKED ISSUE #77925: Installing any libcontainer related software (runc, podman, cri-o) bricks nixos-container
**Describe the bug**
NixOS Containers expect `/etc/containers` to be used by it. but this is a directory that is used by `libcontainer`-compatible systems like podman. More specifically it creates a file `/etc/containers/registries.conf`

`nixos-container.pl` then hard-crashes as it expects its own `*.conf` files to live there on the following line: https://github.com/NixOS/nixpkgs/blob/a0a3f5c86ce04a820035855a081d3acb7f353201/pkgs/tools/virtualization/nixos-container/nixos-container.pl#L153  but that file but it dies instead

**To Reproduce**
Install something `libcontainer`-y and use it
nixos-container is now broken

**Expected behavior**
nixos-container should not namesquat `/etc/containers`

**Screenshots**

**Additional context**

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by adisbladis**
###### Motivation for this change
We need to move NixOS containers somewhere else so these don't clash with Podman, Skopeo & other container software in the libpod & cri-o/cri-u/libcontainer ecosystems.

The state directory move is not strictly a requirement but is good for consistency.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Closes #77925

### CODE REVIEW

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (cole-h) [inline nixos/modules/virtualisation/nixos-containers.nix:12]: Maybe something like this would work?

```suggestion
configurationPrefix = optionalString (versionAtLeast config.system.stateVersion "20.09") "nixos-";
configurationDirectory = "/etc/${configurationPrefix}containers";
stateDirectory = "/var/lib/${configurationPrefix}containers";
```

Though I guess it's less explicit. Just something I thought of while looking at this.

Comment (adisbladis) [reply nixos/modules/virtualisation/nixos-containers.nix]: I like it :) Thanks!

**Comment by flokli** [comment]
This is only really gonna fix podman etc. for all new NixOS installations.

Existing installations will still stay on `/var/lib/containers`, so podman will still not work properly.

I assume we don't want to only be able to provide podman for all new installations… Instead of making this conditional on `stateVersion`, could we ship some "migrate from `/var/lib/containers -> /var/lib/nixos-containers` code for 1-2 releases (if we can clearly detect nixos-container-generated files?)

**Comment by adisbladis** [comment]
> could we ship some "migrate from /var/lib/containers -> /var/lib/nixos-containers code for 1-2 releases

I thought about it but opted not to because:
1. Automatic migrations of user data feels scary
2. They would break rollbacks to previous releases unless we also backported a back-migration

I could add migrations if we are in agreement that it's the correct solution.

**Comment by flokli** [comment]
At least the `/etc/[nixos-]containers` migration should be possible to do independent of `stateVersion`, and fix #77925, should it?

I agree migrating container root directories in `/var/lib/containers` is scary, and indeed shouldn't be done automatically (as it contains all data inside these containers too).

But I'm also not a fan of keeping users upgrading from an older NixOS with state in `/var/lib/containers` indefinitely - we might run into hard-to-debug issues with other container software if things are still left in `/var/lib/containers`.

I'm not sure what's the best approach there, but feel like the first part (new configuration files) should be less problematic and handled first.

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (Ma27) [inline nixos/modules/virtualisation/nixos-containers.nix:9]: Hmm, but since we're not supposed to bump the state-version on a system-update, wouldn't this break 20.03 installations when upgrading to 20.09 if there are existing nixos-containers? To fix this, I'd suggest to raise a warning if `/var/lib/containers` is used.

**Comment by FRidh** [comment]
Do I understand correctly that we were using the generic `/var/lib/containers`, but now because podman et al. also use that generic name, we feel like we should change our container implementation to a different location?

**Comment by Ma27** [comment]
Why did those clash in the first place though? Shouldn't new modules use a better named state-directory from the beginning?

**Comment by adisbladis** [comment]
> Do I understand correctly that we were using the generic /var/lib/containers, but now because podman et al. also use that generic name, we feel like we should change our container implementation to a different location?

Yes. Both `/var/lib/containers` and `/etc/containers` are clashing.

> Why did those clash in the first place though? Shouldn't new modules use a better named state-directory from the beginning?

These locations are often hard-coded and not provided by the NixOS module system.
This becomes extra problematic as it's a whole ecosystem of tools and not just one or two packages that we could patch.
But.. Even if we could patch those tools we would break those tools on non-NixOS so that wouldn't be a great solution even if we could.

The whole situation is very unfortunate but I don't see a way forward except moving NixOS-containers elsewhere.

**Comment by Ma27** [comment]
> This becomes extra problematic as it's a whole ecosystem of tools and not just one or two packages that we could patch

I see. It's fine then :)

> The whole situation is very unfortunate but I don't see a way forward except moving NixOS-containers elsewhere.

Have you checked whether one can migrate NixOS containers to the new location without too much inconveniences? If one has a system with a stateVersion <20.09 and wants to use e.g. `podman`, they'd have to migrate anyway, right?

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (edolstra) [reply nixos/modules/virtualisation/nixos-containers.nix:9]: When you upgrade, you generally don't change `stateVersion`. So existing installations that upgrade to 20.09 would continue to use `/var/lib/containers` and they wouldn't break (unless you use podman or other tools).

Comment (Ma27) [reply nixos/modules/virtualisation/nixos-containers.nix:9]: I'm well-aware of that, but if you deploy a podman instance to an existing system you'll run into issues (as you already mentioned), that's why I brought this up :)

**Comment by edolstra** [comment]
I agree that using `stateVersion` is the most practical solution if we don't want to migrate containers automatically. We could issue a warning if you try to enable nixos-containers *and* podman *and* `stateVersion` < 20.09.

**Comment by worldofpeace** [comment]
@adisbladis News?

**Comment by adisbladis** [comment]
@worldofpeace No news. This is ready for review/merge.

**Review by worldofpeace — APPROVED** [review]
Seems fine to me. Someone more in tune with these components should review it though.

File: nixos/doc/manual/release-notes/rl-2009.xml
Comment (Ma27) [inline nixos/doc/manual/release-notes/rl-2009.xml:314]: `<xref linkend="opt-system.stateVersion" />` is better here since a user can directly navigate to the option's documentation in the rendered releasenotes :)

File: pkgs/tools/virtualization/nixos-container/default.nix
Comment (Ma27) [inline pkgs/tools/virtualization/nixos-container/default.nix:6]: I think that we should use the "new" default here, i.e. `/var/lib/nixos-containers` (and `/etc/nixos-containers`).

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (Ma27) [inline nixos/modules/virtualisation/nixos-containers.nix:722]: I'm very sorry if I forgot about something we discussed in the past, but I'm quite unhappy with this since it's generally a bad idea to bump your system's `stateVersion`. This basically means that each system which was provisioned before 20.09 cannot use `virtualisation.containers`.

Wouldn't it be possible to add e.g. an option `stateDir` for `nixos-containers` which is `mkDefault "/var/lib/containers"` on `stateVersion<20.09` and `mkDefault "/var/lib/nixos-containers"` on `stateVersion<=20.09`?

Then you could warn if `/var/lib/containers` is used and an admin could alter `statePath` as soon as all containers are migrated. If that (optional) migration is done, both tools would be usable on one system.

**Review by Ma27 — COMMENTED** [review]
First of all thanks a lot for taking care of this! :tada: 

I really hope I didn't bring up too much stuff that was discussed in the past. If I forget about this again, please feel free to hit me up on e.g. IRC or Matrix :) 

----

__Note to self__:  before merging I'd like to test a container's migration from `/var/lib/containers` to `/var/lib/nixos-containers` just to make sure there are no edge-cases to address.

**Comment by szlend** [comment]
@adisbladis Is there any chance of this still getting into 20.09? The changes required from the code review seem pretty minor. Let me know if there's anything I can do to help, e.g. help test the migration process.

**Comment by Ma27** [comment]
As stated previously, I'd like to see the state-dir being more configurable to make sure that I don't have to bump the state-version of an existing system when I want to migrate my containers to the new location - apart from the other comments I left (/cc @adisbladis).

Optionally, I'd add a warning to the documentation about the current state.

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (adisbladis) [reply nixos/modules/virtualisation/nixos-containers.nix:722]: > I'm very sorry if I forgot about something we discussed in the past, but I'm quite unhappy with this since it's generally a bad idea to bump your system's

This PR doesn't introduce this incompatibility, the incompatibility _already_ exists. What I'm doing right there is making users aware of a dangerous configuration that is already unsupported.

I'm reluctant to add more options at the top-level of this module, especially as it doesn't have any namespacing, but instead introduces new arguments in the `boot` attribute.

**Comment by critbase** [comment]
Still matters to me

**Comment by adisbladis** [comment]
Years later and this is still relevant..

I have rebased this, moved the release notes to 22.05 & fixed up some breakage.
I think this is ready and should go in 22.05.

**Comment by flokli** [comment]
I ran all container-related tests, and gave this another look. LGTM, merged!

**Pull request merged**
